### PR TITLE
[2612] Update control of rng seeds

### LINKED ISSUE #2612: Reproducible performance benchmarks
### Describe the task. It can be a feature, documentation, etc.

In recent performance experiments, we have seen significant variability in peak memory using the same configuration, similarly for throughput results. 
Therefore we need best practices for assessing performance.

Requirements:
- reproducibility: ensure fixed seeds (especially in performance-relevant components like data sampling).
- representative sample size: to ensure that the variability in hardware and the variability in data is sufficiently averaged out (how many samples are required to attain which level of confidence?)
- (potentially) representative samples: instead of relying on a large sample size capturing the data distribution (especially in case of a fixed seed, we might get unlucky and the sampled data for a given training duration is not representative of the training jobs we are optimizing for), we the data could be chosen to reflect the training distribution.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

https://gitlab.jsc.fz-juelich.de/hedgedoc/hXo2AzgjTfCa4O0wEdOUgQ?both

### Area

- [x] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [x] performance

### PR DESCRIPTION
**Pull request by florianscheidl**
## Description

RNG seeds were missing or hard-coded to a function of time, precluding reproducible experiments, important for performance benchmarking.

We ensure the seed is configurable in the config and propagates to
- torch, numpy, Python's random seed
- spoof data
- across GPUs through an all-reduce.

We add tests to verify the reproducibility in the affected paths.

## Issue Number

Fixes #2612.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/datasets/stream_data_test.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# (C) Copyright 2025 WeatherGenerator contributors.
```
Comment (clessig) [inline src/weathergen/datasets/stream_data_test.py]: We don't use the tests so I don't think we will merge

File: src/weathergen/train/trainer_base.py [hunk 2]
```diff
@@ -62,12 +65,34 @@ def init_torch(use_cuda=True, num_accs_per_task=1, multiprocessing_method="fork"
...
+        random.seed(seed)
+        np.random.seed(seed % 2**32)
+
     @staticmethod
     def init_ddp(cf):
         """Initializes the distributed environment."""
         rank = 0
         local_rank = 0
 
+        # rng seed: use value from config if provided, otherwise derive from time;
```
Comment (clessig) [inline src/weathergen/train/trainer_base.py:78]: Why is this in `init_ddp()`? I think it's ok all things considered, but it's not super natural.

File: src/weathergen/train/trainer_base.py [hunk 3]
```diff
@@ -127,14 +152,16 @@ def init_ddp(cf):
...
+            # communicate rank 0's rng seed to all ranks; ranks derive their own seed
+            # from a common base seed (cf. MultiStreamDataSampler.worker_workset)
+            l_seed = torch.tensor(
+                [cf.data_loading.rng_seed if rank == 0 else 0], dtype=torch.int64
+            ).to(device)
+            dist.all_reduce(l_seed, op=torch.distributed.ReduceOp.SUM)
+            cf.data_loading.rng_seed = l_seed.item()
+
+        TrainerBase.init_seeds(cf.data_loading.rng_seed)
+        print(f"rank: {rank} uses rng_seed: {cf.data_loading.rng_seed}")
```
Comment (clessig) [inline src/weathergen/train/trainer_base.py]: Remove

File: src/weathergen/train/trainer_base_test.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+# (C) Copyright 2025 WeatherGenerator contributors.
```
Comment (clessig) [inline src/weathergen/train/trainer_base_test.py]: Same as above: we don't have a mechanism to run these tests so we shouldn't merge

**Review by clessig — COMMENTED** [review]
Looks good, just some minor comments.

File: src/weathergen/train/trainer_base.py (hunk 2 above)
Comment (florianscheidl) [reply src/weathergen/train/trainer_base.py:78]: You have a good point. How about we route it through init_seeds?

File: src/weathergen/datasets/stream_data_test.py (hunk 1 above)
Comment (florianscheidl) [reply src/weathergen/datasets/stream_data_test.py:1]: ./scripts/actions.sh unit-test starts this one and the one below.
